core/src/server/rpc_module.rs

```diff
@@ -777,47 +787,42 @@ impl SubscriptionSink {
 		S: Stream<Item = T> + Unpin,
 		T: Serialize,
 	{
-		let mut close_stream = self.close.clone();
-		let mut item = stream.next();
-		let mut close = close_stream.next();
-
-		loop {
-			match futures_util::future::select(item, close).await {
-				Either::Left((Some(result), c)) => {
-					match self.send(&result) {
-						Ok(_) => (),
-						Err(Error::SubscriptionClosed(close_reason)) => {
-							self.close(&close_reason);
-							break Ok(());
-						}
-						Err(err) => {
-							tracing::error!("subscription `{}` failed to send item got error: {:?}", self.method, err);
-							break Err(err);
-						}
-					};
-					close = c;
-					item = stream.next();
-				}
-				// No messages should be sent over this channel
-				// if that occurred just ignore and continue.
-				Either::Right((Some(_), i)) => {
-					item = i;
-					close = close_stream.next();
-				}
-				// Connection terminated.
-				Either::Right((None, _)) => {
-					self.close(&SubscriptionClosed::new(SubscriptionClosedReason::ConnectionReset));
-					break Ok(());
+		if let Some(close_notify) = self.close_notify.clone() {
+			let mut item = stream.next();
+			loop {
+				match futures_util::future::select(item, Box::pin(close_notify.notified())).await {
```

Comment:
To save some allocatin,g how about moving the `Box::pin(close_notify.notified())` up next to the `let mut item`, and "re-set" it on each loop like we do the item on line 803?

Reply:
(also I wonder whether we could use https://docs.rs/futures/latest/futures/macro.pin_mut.html instead of `Box::pin`?)

Reply:
good catch missed that